src/OmniSharp.Roslyn.CSharp/Workers/Diagnostics/CSharpDiagnosticService.cs

```diff
@@ -1,152 +1,222 @@
             }
+        }
 
-            Task.Factory.StartNew(async () =>
-            {
-                await Task.Delay(100);
-                await Dequeue();
+        private IEnumerable<Task> WaitForPendingWorkIfNeededAndGetIt(IEnumerable<ProjectId> projectIds)
+        {
+            return _workQueue
+                .Where(x => projectIds.Any(pid => pid == x.Key))
+                .Select(x => Task.Delay(10 * 1000, x.Value.workReadySource.Token)
+                    .ContinueWith(task => LogTimeouts(task, x.Key.ToString())));
+        }
 
-                if (_openDocuments.IsEmpty)
-                {
-                    lock (_lock)
-                    {
-                        _queueRunning = false;
-                    }
-                }
-                else
-                {
-                    this.ProcessQueue();
-                }
-            });
+        private Task WaitForInitialStartupWorkIfAny()
+        {
+            return Task.Delay(10 * 1000, _initializationQueueDoneSource.Token)
+                                .ContinueWith(task => LogTimeouts(task, nameof(_initializationQueueDoneSource)));
+        }
+
+        private void LogTimeouts(Task task, string description)
+        {
+            if (!task.IsCanceled) _logger.LogError($"Timeout before work got ready for {description}.");
+        }
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs changeEvent)
+        {
+            if (changeEvent.Kind == WorkspaceChangeKind.DocumentChanged
+                || changeEvent.Kind == WorkspaceChangeKind.DocumentAdded
+                || changeEvent.Kind == WorkspaceChangeKind.ProjectAdded)
+            {
+                var project = changeEvent.NewSolution.GetProject(changeEvent.ProjectId);
+                QueueForAnalysis(new[] { project });
+            }
         }
 
-        private async Task Dequeue()
+        private async Task Analyze(Project project, CancellationTokenSource workReadySource, CancellationToken token)
         {
-            var tasks = new List<Task<DiagnosticResult>>();
-            for (var i = 0; i < 50; i++)
+            try
             {
-                if (_openDocuments.IsEmpty)
+                // Only basic syntax check is available if file is miscellanous like orphan .cs file.
+                // Todo: Where this magic string should be moved?
+                if (project.Name == "MiscellaneousFiles")
                 {
-                    break;
+                    await AnalyzeSingleMiscFilesProject(project);
+                    return;
                 }
 
-                if (_openDocuments.TryDequeue(out var filePath))
+                var allAnalyzers = _providers
+                    .SelectMany(x => x.CodeDiagnosticAnalyzerProviders)
+                    .Concat(project.AnalyzerReferences.SelectMany(x => x.GetAnalyzersForAllLanguages()));
```

Comment:
there is method that return per lanaguage analyzer? the method has Project given so you can get project.languagename?